Mix: fix import_config with non-matching wildcard

oneeman opened the pull request:
Closes #3120

I'm not entirely sure about this.  It does fix the problem, while keeping the behavior of raising an error if a non-glob name of a non-existent file is received (as tested in `"import_config/1 with bad path"`).  But it seems less than ideal that a function called `read_wildcard!` handles the degenerate case of a concrete filename with no wildcards as a special case.  It would make more sense for it to treat it as a non-matching glob too (just using `Path.wildcard(path)` regardless of whether or not there are any results), but that would break the functionality of raising an error if called with a non-glob filename for a file that doesn't exist.  

There are other ways one could go (e.g. to modify `import_config` so it call either `read!` or `read_wildcard!` depending on whether or not it's a glob), but since these are public functions I figured it's best to start with the smallest change and get some feedback.

ericmj commented:
Wouldn't it be possible to just call `Path.wildcard/1` with no case or regex? That should give us get correct behaviour.

```iex
iex(1)> File.ls!
["config.exs", "dev.exs", "prod.exs", "test.exs"]
iex(2)> Path.wildcard("*")
["config.exs", "dev.exs", "prod.exs", "test.exs"]
iex(3)> Path.wildcard("config.exs")
["config.exs"]

oneeman commented:
@ericmj That was what I tried at first, but it breaks the test at https://github.com/oneeman/elixir/blob/mix-config-bugfix/lib/mix/test/mix/config_test.exs#L70-L76.  It is an option to consider, though.

josevalim commented on lib/mix/lib/mix/config.ex:
```diff
@@ -146,9 +146,10 @@ defmodule Mix.Config do
-    paths = case Path.wildcard(path) do
-      [] -> [path]
-      o  -> o
+    paths = if path =~ ~r/[*?[{]/ do
```
Maybe `String.contains?(path, ~w(* ? {))`?

josevalim replied:
Also, should we add some code comments here explaining that we are checking for the path because we would like to raise for missing files?

josevalim commented:
I ahve dropped some comments too! Thanks for tackling this!

oneeman commented:
Thank you, that's perfect, I hadn't realized that `String.contains?` also accepts a list.  I added a commit with it.  For the comment, I added the note to the documentation, to make that case clear both to users looking at the API and anyone looking at the source code.  If this is all ok let me know and I'll squash the commits.

josevalim commented:
Perfect, let's squash and merge!  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

oneeman commented:
Ready to go :-)

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
